explainability-service/src/main/java/org/kie/trustyai/service/data/utils/UploadUtils.java

```diff
@@ -86,24 +99,134 @@ private static Object[][] transpose(Object[] objectList) {
+            }
+        }
+
+        // make sure all column names are unique
+        List<String> errors = new ArrayList<>();
+        if (names.size() != payload.getTensorPayloads().length) {
+            errors.add(String.format(
+                    "Each %s tensor must have unique names. However, the following duplicate names were found: %s",
+                    contentType, nameDups));
+        }
+
+        // make sure all columns have matched length
+        if (firstDims.size() > 1) {
+            errors.add(String.format(
+                    "Each %s tensor must have the same sized first dimension (e.g., shape[0]). However, the following first dimensions were passed: %s",
+                    contentType, firstDims));
+        }
+
+        // assert that all columns are vectors
+        if (laterDims.size() != 1 && !laterDims.contains(1)) {
+            errors.add(String.format(
+                    "Higher dimensional %ss (e.g., shape=[m, n] where n>1) are not yet supported.", contentType));
+        }
+
+        // report all validation errors together
+        if (!errors.isEmpty()) {
+            StringBuilder errorMsg = new StringBuilder();
+            errorMsg.append(String.format("One or more errors were found with the inbound %s payload:", payloadType));
+            for (String error : errors) {
+                errorMsg.append(System.lineSeparator() + " - ");
+                errorMsg.append(error);
+            }
+            throw new IllegalArgumentException(errorMsg.toString());
+        }
+    }
+
+    private static <T, U extends ModelInferBasePayload> List<T> getBuilderInputsOrOutputs(U payload, String payloadType, String contentType, Function<TensorContructionHolder, T> tensorCreator) {
+        List<T> builtPayloads = new ArrayList<>();
+
+        // if we have multiple tensors, assume each tensor describes a data column
+        if (payload.getTensorPayloads().length > 1) {
+            validateMultiTensors(payload, payloadType, contentType);
+
+            for (TensorPayload tp : payload.getTensorPayloads()) {
+                Object[] vector;
+
+                // if we have data where shape=[1, x], values=[[a], [b], [c]], transpose and flatten to shape=[x], values=[a,b,c]
+                if (tp.getData()[0] instanceof List) {
+                    vector = transpose(tp.getData())[0];
+                    Number[] reverseShape = tp.getShape().clone();
+                    ArrayUtils.reverse(reverseShape);
+                    tp.setShape(reverseShape);
+                } else {
+                    vector = tp.getData();
+                }
+                builtPayloads.add(tensorCreator.apply(new TensorContructionHolder(tp, vector)));
+
+            }
+        } else { //otherwise, assume that we have one tensor, holding data in row vector(s)
+            TensorPayload tp = payload.getTensorPayloads()[0];
+
+            // is the payload a matrix/tensor?
+            if (tp.getData()[0] instanceof ArrayList) {
```

Comment:
Here we check for being a class (`ArrayList`), above we check for being an interface (`List`).
I would prefer to check both for the interface, but at least both consistent (e.g. could be both `ArrayList` too, if we're sure that's the class)